File: dart/blockchain/lib/consensus/consensus.dart
import 'package:blockchain/codecs.dart';
import 'package:blockchain/common/block_height_tree.dart';
import 'package:blockchain/common/clock.dart';
import 'package:blockchain/common/parent_child_tree.dart';
import 'package:blockchain/common/resource.dart';
import 'package:blockchain/consensus/block_header_validation.dart';
import 'package:blockchain/consensus/chain_selection.dart';
import 'package:blockchain/consensus/staker_tracker.dart';
import 'package:blockchain/consensus/eta_calculation.dart';
import 'package:blockchain/consensus/leader_election_validation.dart';
import 'package:blockchain/consensus/local_chain.dart';
import 'package:blockchain/consensus/models/protocol_settings.dart';
import 'package:blockchain/crypto/utils.dart';
import 'package:blockchain/data_stores.dart';
import 'package:blockchain_protobuf/models/core.pb.dart';
import 'package:ribs_core/ribs_core.dart';
import 'package:ribs_effect/ribs_effect.dart';

class Consensus {
  final BlockHeaderValidation blockHeaderValidation;
  final ChainSelection chainSelection;
  final StakerTracker stakerTracker;
  final EtaCalculation etaCalculation;
  final LeaderElection leaderElection;
  final LocalChain localChain;

  Consensus({
    required this.blockHeaderValidation,
    required this.chainSelection,
    required this.stakerTracker,
    required this.etaCalculation,
    required this.leaderElection,
    required this.localChain,
  });

  static Resource<Consensus> make(
          ProtocolSettings protocolSettings,
          DataStores dataStores,
          Clock clock,
          FullBlock genesisBlock,
          CurrentEventIdGetterSetters currentEventIdGetterSetters,
          ParentChildTree<BlockId> parentChildTree,
          BlockHeightTree blockHeightTree,
          DComputeImpl isolate) =>
      Resource.eval(IO.fromFutureF(() async {
        final genesisBlockId = genesisBlock.header.id;
        final etaCalculation = EtaCalculationImpl(dataStores.headers.getOrRaise,
            clock, genesisBlock.header.eligibilityCertificate.eta.decodeBase58);

        final leaderElection = LeaderElectionImpl(protocolSettings, isolate);

        final epochBoundaryState = epochBoundariesEventSourcedState(
          clock,
          await currentEventIdGetterSetters.epochBoundaries.get(),
          parentChildTree,
          currentEventIdGetterSetters.epochBoundaries.set,
          dataStores.epochBoundaries,
          dataStores.headers.getOrRaise,
        );
        final consensusDataState = ConsensusData.eventSourcedState(
            await currentEventIdGetterSetters.consensusData.get(),
            parentChildTree,
            currentEventIdGetterSetters.consensusData.set,
            ConsensusData(
                dataStores.delayedActiveStake,
                dataStores.delayedInactiveStake,
                dataStores.delayedActiveStakers),
            dataStores.bodies.getOrRaise,
            dataStores.transactions.getOrRaise);

        final stakerTracker = StakerTrackerImpl(
            genesisBlockId, epochBoundaryState, consensusDataState, clock);

        return LocalChainImpl.make(
                genesisBlockId,
                await currentEventIdGetterSetters.canonicalHead.get(),
                blockHeightTree,
                (id) async => (await dataStores.headers.getOrRaise(id)).height,
                parentChildTree)
            .flatTap((localChain) =>
                stakerTracker.epochBoundaryState.followChain(localChain))
            .flatTap((localChain) => ResourceUtils.forStreamSubscription(() =>
                localChain.adoptions
                    .asyncMap(currentEventIdGetterSetters.canonicalHead.set)
                    .listen(null)))
            .map((localChain) {
          final chainSelection =
              ChainSelection(protocolSettings: protocolSettings);
          final headerValidation = BlockHeaderValidationImpl(
              genesisBlockId,
              etaCalculation,
              stakerTracker,
              leaderElection,
              clock,
              dataStores.headers.getOrRaise);

          return Consensus(
              blockHeaderValidation: headerValidation,
              chainSelection: chainSelection,
              stakerTracker: stakerTracker,
              etaCalculation: etaCalculation,
              leaderElection: leaderElection,
              localChain: localChain);
        });
      })).flatMap(identity);
}

File: scala/core/src/main/scala/blockchain/consensus/Consensus.scala
package blockchain.consensus

import blockchain.*
import blockchain.codecs.given
import blockchain.crypto.CryptoResources
import blockchain.models.*
import blockchain.utility.{Exp, Log1P}
import cats.effect.{Async, Resource}
import cats.implicits.*
import cats.effect.implicits.*

case class Consensus[F[_]](
    headerValidation: HeaderValidation[F],
    chainSelection: ChainSelection[F],
    stakerTracker: StakerTracker[F],
    etaCalculation: EtaCalculation[F],
    leaderElection: LeaderElection[F],
    localChain: LocalChain[F]
)

object Consensus:
  def make[F[_]: Async](
      genesis: FullBlock,
      clock: Clock[F],
      dataStores: DataStores[F],
      cryptoResources: CryptoResources[F],
      eventIdGetterSetters: EventIdGetterSetters[F],
      blockIdTree: BlockIdTree[F],
      blockHeights: BlockHeights.BSS[F]
  ): Resource[F, Consensus[F]] =
    for {
      protocolSettings <- Resource.pure[F, ProtocolSettings](
        ProtocolSettings.Default.merge(genesis.header.settings)
      )
      etaCalculation <- EtaCalculation.make[F](
        dataStores.headers.getOrRaise,
        clock,
        genesis.header.eligibilityCertificate.eta,
        cryptoResources.blake2b256,
        cryptoResources.blake2b512,
        cryptoResources.ed25519VRF
      )
      epochBoundariesBSS <- EpochBoundaries.make[F](
        dataStores.epochBoundaries.pure[F],
        eventIdGetterSetters.epochBoundaries.get(),
        blockIdTree,
        eventIdGetterSetters.epochBoundaries.set,
        clock,
        dataStores.headers.getOrRaise
      )
      stakerDataBSS <- StakerData.make[F](
        StakerData
          .State(
            dataStores.activeStake,
            dataStores.inactiveStake,
            dataStores.stakers
          )
          .pure[F],
        eventIdGetterSetters.stakerData.get(),
        blockIdTree,
        eventIdGetterSetters.stakerData.set,
        dataStores.bodies.getOrRaise,
        dataStores.transactions.getOrRaise
      )
      stakerTracker <- StakerTracker
        .make[F](clock, genesis.header.id, stakerDataBSS, epochBoundariesBSS)
      canonicalHeadId <- eventIdGetterSetters.canonicalHead.get().toResource
      localChain <- LocalChain
        .make[F](genesis.header.id, blockHeights, canonicalHeadId)
      chainSelection <- ChainSelection.make[F](
        cryptoResources.blake2b512,
        cryptoResources.ed25519VRF,
        protocolSettings.chainSelectionKLookback,
        protocolSettings.chainSelectionSWindow
      )
      exp <- Exp.make()
      log1p <- Log1P.make() >>= Log1P.makeCached
      leaderElection <- LeaderElection.make(
        protocolSettings,
        cryptoResources.blake2b512,
        exp,
        log1p
      )
      headerValidation <- HeaderValidation.make[F](
        genesis.header.id,
        etaCalculation,
        stakerTracker,
        leaderElection,
        clock,
        dataStores.headers.getOrRaise,
        cryptoResources
      )
    } yield Consensus(
      headerValidation,
      chainSelection,
      stakerTracker,
      etaCalculation,
      leaderElection,
      localChain
    )
